app-cpp/appPtr.hpp

```diff
@@ -0,0 +1,436 @@
+    }
+
+    // Copy Operand
+    shared_ptr& operator=(const shared_ptr& inCopy)
+    {
+        // Copying to self should be a NOP
+        if (this != &inCopy)
+        {
+            mData = inCopy.mData;
+            mControl = inCopy.mControl;
+            mControl->IncrementStrong();
+        }
+        return *this;
+    }
+
+    T* operator->()
+    {
+        return get();
+    }
+
+    T& operator*()
+    {
+        return *get();
+    }
+
+    operator bool() const
+    {
+        return (mData != nullptr);
+    }
+
+    T* get()
+    {
+        return mData;
+    }
+
+private:
+    class ControlBlock
+    {
+    public:
+        ControlBlock(T* inData) :  
+                mData{inData}
+        {
+            // Do nothing
+        }
+
+        // RO5 Methods added
+        ~ControlBlock() = default;
+
+        T* get()
+        {
+            return mData;
+        }
+
+        std::pair<std::size_t, std::size_t> IncrementStrong()
+        {
+            std::lock_guard<std::mutex> lock{mMutex};
+            auto refCounts = std::make_pair(++mStrongCount, mWeakCount);
+            return refCounts;
+        }
+
+        std::pair<std::size_t, std::size_t> DecrementStrong()
+        {
+            std::lock_guard<std::mutex> lock{mMutex};
+            auto refCounts = std::make_pair(--mStrongCount, mWeakCount);
+            return refCounts;
+        }
+
+        std::pair<std::size_t, std::size_t> IncrementWeak()
+        {
+            std::lock_guard<std::mutex> lock{mMutex};
+            auto refCounts = std::make_pair(mStrongCount, ++mWeakCount);
+            return refCounts;
+        }
+
+        std::pair<std::size_t, std::size_t> DecrementWeak()
+        {
+            std::lock_guard<std::mutex> lock{mMutex};
+            auto refCounts = std::make_pair(mStrongCount, --mWeakCount);
+            return refCounts;
+        }
+
+    private:
+        std::size_t mStrongCount{1}; // Default constructs with initial refcount of 1
+        std::size_t mWeakCount{0};
+        std::mutex mMutex{};
+        T* mData{}; // Used by weak pointers to obtain a new shared pointer
+    };              // class ControlBlock
+
+private:
+    template <typename T>
+    friend class weak_ptr;
+
+    shared_ptr(ControlBlock* inBlock)
+    {
+        mData = inBlock->get();
+        mControl = inBlock;
+        mControl->IncrementStrong();
+    }
+
+private:
+    T* mData{};
+    ControlBlock* mControl{};
+}; // class shared_ptr<T>
+
+// TRICKY: Variatic template arguments 
+// Template that takes any number and type of arguments
+template<typename T, typename ...Args>
+shared_ptr<T> make_unique(Args... inArgs)
+{
+    T* data = new T{std::forward<Args>(inArgs)...};
+    shared_ptr<T> unique{data};
+    return unique;
+}
+
+// WEAK POINTER
+template <typename T>
+class weak_ptr
+{
+public:
+    weak_ptr() : 
+        mControl{nullptr}
+    {
+
+    }
+
+    weak_ptr(shared_ptr<T>& inPtr)
+    {
+        if (inPtr)
```

Comment:
I think this should be `if (inPtr->mControl != nullptr)`...
Because you want to use mControl, not mData...
and someone could pass you an empty shared_ptr